aslam_cv_pipeline/src/visual-npipeline.cc

```diff
@@ -61,11 +61,14 @@ bool VisualNPipeline::processImageBlockingIfFull(size_t camera_index, const cv::
     std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
IMHO this can create a race condition, namely in this case:
thread A comes along, checks the while condition, shutdown_ is false, enters the while loop (but doesn't acquire the lock just yet), 
not thread B comes along, calls shutdown(), acquires the lock, sets shutdown_ to true, calls notifyAll(), 
not thread A comes back to live, acquires the lock, completed_ is full, so it waits... and never ever ever evvvveeeeer wakes up again :-)
that's why I think the lock needs to be acquired here outside and before the while loop.

Reply:
Technically, thread A would return with false (see line 64). But you are right, stuff might happen between checking shutdown and acquiring the lock. I think it makes sense to move it outside the while loop. At least I see no harm.

Reply:
By doing that, I actually see no argument for using an atomic shutdown_ if we just lock before reading from it or writing to it. Agreed?